### DISCUSSION TITLE
MQTT Discovery based integration?

### DISCUSSION THREAD

**Discussion by tomdavidson** [zwave-js/node-red-contrib-zwave-js#287] [Q&A] [answered]
Can we use a MQTT broker in parallel with Node-Red?  Similar to the Home Assistant integration that uses MQTT discovery rather than web-socket connection?

I'd like to have the tight integration that node-red-contrib-zwave-js offers but I also want other software interact with the z-wave network and use non-zwave networks with Node-Red, such as with zigbee2mqtt.

**Comment by marcus-j-davies** [comment]
Hi @tomdavidson 

`node-red-contrib-zwave-js` is purely a Z-Wave Runtime, running directly in Node RED.
Node RED its-self has MQTT nodes (IN/OUT)

This module itself, isn't equipped with MQTT features, it's not something that I plan on adding.

But there is nothing stopping you using MQTT IN/OUT nodes to alter the state of Z-Wave Network running in Node RED, or for the Z-Wave Network, to forward events via MQTT

Example

![Screenshot 2023-12-02 at 10 47 33](https://github.com/zwave-js/node-red-contrib-zwave-js/assets/55892693/f036b954-a14a-40e6-9f56-a12dbbd9ba01)

You will need  to maybe format the the payloads, before forwarding/processing

**Comment by tomdavidson** [comment]
Thank you for the quick reply. 

> running directly in Node RED.
So if node-red needs restarted then the zwave network needs reassembled? Not necessarily wrong for my application, just making sure I understand.

Using Node-Red as the MQTT proxy does seem simple (to one that only has 1/2 a clue such as I) but I am worried about "re-inventing" zwave-js-ui. The other option I was looking at is zwave-js-ui (and zigbee2mqqt) in parallel to Node-Red.

Do you have any other insights contrasting the two paths that you can share?

**Comment by marcus-j-davies** [comment, marked as answer]
> So if node-red needs restarted then the zwave network needs reassembled? Not necessarily wrong for my application, just making sure I understand.

No - `node-red-contrib-zwave-js` uses the same driver as all other Z-Wave JS implementations, meaning all network cache data is written to disc and restored when ever the runtime is started up again (nothing needs to get re-interviewed after a restart) - provided Node RED is using a writable file system. - and of course , device interviews are complete.

I store all network data in `~/.node-red/zwave-js-cache` on a normal Node RED install

`zwave-js-ui` - is an MQTT implementation of the same driver. `zwave-js-ui ` is  @robertsLando's baby  - we are both on the same Team - and we all develop different implementations for different platforms.

ZWave JS is implemented in various shapes.

- A Native Node RED Module (what your discussing now) - Mine
- zwave-js-ui An MQTT orientated implementation
- An IO.Broker Module (developed by @AlCalzone - the same wizard who actually developed Z-Wave JS 🤓 )
- Zwave JS Server  (a Web-socket implementation)- to provide the runtime for Home Assistant (given its python)
- ZWaveJS.NET - A wrapper around the library based on the server for c#  - Mine

All implementations provide the same functionality - just in different ways.

My advice is - if you depend heavily on MQTT = `zwave-js-ui`
If you want to manage it in Node RED = `node-red-contrib-zwave-js`
(but you need to polly fill the MQTT parts using the MQTT nodes)

Both offer UI's and all the same API - again just in different forms
